src/spotify_to_tidal/sync.py

```diff
@@ -82,30 +90,48 @@ def get_spotify_artists(spotify_track: t_spotify.SpotifyTrack, do_normalize=Fals
         return True
     return get_tidal_artists(tidal_track, True).intersection(get_spotify_artists(spotify_track, True)) != set()
 
-def match(tidal_track, spotify_track) -> bool:
+def match(tidal_track: tidalapi.Track, spotify_track: t_spotify.SpotifyTrack) -> bool:
     return isrc_match(tidal_track, spotify_track) or (
         duration_match(tidal_track, spotify_track)
         and name_match(tidal_track, spotify_track)
         and artist_match(tidal_track, spotify_track)
     )
 
+# Function for caching purposes
+@cached(TTLCache(maxsize=50, ttl=300))
+def get_album_tracks(album: tidalapi.Album) -> Sequence[tidalapi.Track]:
+    return album.tracks()
 
 def tidal_search(spotify_track_and_cache, tidal_session: tidalapi.Session) -> tidalapi.Track | None:
     spotify_track, cached_tidal_track = spotify_track_and_cache
-    if cached_tidal_track: return cached_tidal_track
+    if cached_tidal_track:
+        return cached_tidal_track
     # search for album name and first album artist
-    if 'album' in spotify_track and 'artists' in spotify_track['album'] and len(spotify_track['album']['artists']):
-        album_result = tidal_session.search(simple(spotify_track['album']['name']) + " " + simple(spotify_track['album']['artists'][0]['name']), models=[tidalapi.album.Album])
-        for album in album_result['albums']:
-            album_tracks = album.tracks()
-            if len(album_tracks) >= spotify_track['track_number']:
-                track = album_tracks[spotify_track['track_number'] - 1]
-                if match(track, spotify_track):
-                    return track
+    if (
+        "album" in spotify_track
+        and "artists" in spotify_track["album"]
+        and len(spotify_track["album"]["artists"])
+    ):
+        for artist in spotify_track["album"]["artists"]:
+            artist_name = artist["name"]
+            album_name = spotify_track["album"]["name"]
+            album_result = search_tidal_albums(tidal_session, artist_name, album_name)
+            for album in album_result["albums"]:
+                album_tracks = get_album_tracks(album)
+                if len(album_tracks) >= spotify_track["track_number"]:
+                    track = album_tracks[spotify_track["track_number"] - 1]
+                    if match(track, spotify_track):
+                        TidalTrackCache().data[spotify_track['id']] = track.id
+                        return track
     # if that fails then search for track name and first artist
-    for track in tidal_session.search(simple(spotify_track['name']) + ' ' + simple(spotify_track['artists'][0]['name']), models=[tidalapi.media.Track])['tracks']:
-        if match(track, spotify_track):
-            return track
+    spotify_track_name = spotify_track["name"]
+    for artist in spotify_track["artists"]:
```

Comment:
It looks like you've brought in an algorithm change as well here right? Sorry to  be difficult, but I'd really prefer to discuss any search algorithm changes separately from the caching. Would you be able to split anything not related to caching into a separate PR? As mentioned in the other PR I'd want to see some data (unit tests would be even better if possible but at least some real-world data) proving that there's an improvement in the match rate